### DISCUSSION TITLE
zh/modes/val/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#10401] [Docs]
# zh/modes/val/

YOLOv8 模型验证指南。通过Python 和CLI 示例，了解如何使用验证设置和指标来评估YOLO 模型的性能。

https://docs.ultralytics.com/zh/modes/val/

**Comment by Ameridru** [comment]
请问在模型验证中，计算Precision、Recall时默认采用的IOU阈值是0.5吗，能否打印出不同IOU阈值下的Precision和Recall？

**Reply by glenn-jocher** [reply]
在 YOLOv8 模型验证过程中，默认使用的 IOU 阈值确实是 0.5 (`iou=0.5`)，用于计算 Precision 和 Recall。如果您想查看不同 IOU 阈值下的 Precision 和 Recall，可以通过修改 `iou` 参数值来实现。不过，请注意，目前 YOLOv8 验证模式的输出直接在 CLI 或 Python API 中不提供不同 IOU 阈值的详细数据。您可能需要自定义处理结果数据或使用其他工具来实现这一需求。

在 Python 中，您可以这样设置不同的 IOU 阈值：

```python
from ultralytics import YOLO

# Load the model
model = YOLO('yolov8n.pt')

# Validate the model with a custom IOU threshold
iou_values = [0.5, 0.6, 0.7]  # Example IOU values
for iou in iou_values:
    metrics = model.val(iou=iou)
    print(f"IOU: {iou}, mAP: {metrics.box.map50}")
```

请根据您的实际需要自由调整 IOU 值列表 `iou_values`。希望这可以帮助您！如果有更具体的需求或进一步的问题，请随时询问。🙂

**Reply by Ameridru** [reply]
当我使用RTDETR.val()方法的时候，我尝试使用不同的conf参数和iou参数，发现最后输出的性能指标数值不会变化。请问RTDETR是否还不能输出不同的conf和iou下的性能指标？

**Reply by glenn-jocher** [reply]
RT-DETR currently does not support outputting performance metrics for different `conf` and `iou` parameters. You may need to manually adjust these settings and evaluate the results separately. For more detailed insights, please refer to our documentation.

**Comment by echo385** [comment]


**Reply by glenn-jocher** [reply]
@echo385 目前，YOLOv8 的 Val 模式并不直接支持与 sahi 的集成来在高分辨率图像上绘制评价曲线。但您可以在模型验证后独立使用 sahi 或其他可视化工具来处理和展示您的评价曲线。

例如，您可以在验证模型后，使用 Python 进行结果处理并利用 matplotlib 或其他库来绘图。这里是一个基本的代码示例：
```python
from ultralytics import YOLO
import matplotlib.pyplot as plt

# Load and validate the model
model = YOLO('path/to/your/model.pt')
metrics = model.val()  # Make sure the model has training settings integrated

# Example to plot mAP (modify as per your metric requirements)
plt.figure(figsize=(10, 5))
plt.plot(metrics.history['mAP'])
plt.title('Model mAP Over Validation Dataset')
plt.xlabel('Epochs')
plt.ylabel('mAP')
plt.grid(True)
plt.show()
```
这种方法使您可以灵活地展示所需的任何评价指标。希望这能帮助到您！如果有更具体的集成方式或需求，期待您的进一步说明。

**Comment by echo385** [comment]


**Reply by pderrenger** [reply]
@echo385 您好陈香枝，

感谢您的反馈和详细说明！目前YOLOv8的Val模式确实默认要求输入尺寸为640进行验证。如果您希望在不分割图片的情况下直接使用高分辨率图片进行验证，您可能需要对原有代码进行一些自定义修改以适应大尺寸的输入。

一种可能的方法是修改数据加载和预处理步骤，让模型能够接受和处理原始分辨率的图像。这可能需要较深的技术背景和一定的代码修改工作。

如果您需要进一步的具体技术指导，请随时告知，我们乐意提供更多帮助！同时，期待您分享任何成功的适应方案或进一步的问题。

祝好，
Ultralytics 团队 😊

**Comment by echo385** [comment]
请问在验证过程中哪里可以看见目标被识别为背景的图片，我想知道漏检的图片到底是哪些？

**Reply by glenn-jocher** [reply]
To view images where objects were identified as background or missed detections during the validation process, you can enable the `save_json` and `plots` arguments during validation. This will save the results in a JSON file and generate plots that visually compare predictions with ground truths, helping you identify which images contain missed detections.

Here's how you can set it up:

```bash
yolo detect val model=yolov8n.pt data=coco128.yaml save_json=True plots=True
```

This command will save detailed detection results and visualizations, allowing you to review the images and understand the model's performance better.

**Comment by Stephen-Hao** [comment]
类似@Ameridru的问题，在模型验证中，计算Precision、Recall时我可以设定不同的置信度阈值吗，能否打印出不同置信度阈值下的Precision和Recall？

**Reply by glenn-jocher** [reply]
@Stephen-Hao hello,

Thank you for your question! Yes, you can set different confidence thresholds to calculate Precision and Recall during model validation in YOLOv8. This can be done by adjusting the `conf` parameter when running the validation. Here's how you can do it:

### Using Python:
```python
from ultralytics import YOLO

# Load your model
model = YOLO("yolov8n.pt")  # or your custom model path

# Validate the model with different confidence thresholds
for conf_threshold in [0.1, 0.25, 0.5, 0.75]:
    metrics = model.val(conf=conf_threshold)
    print(f"Confidence Threshold: {conf_threshold}")
    print(f"Precision: {metrics.box.precision}")
    print(f"Recall: {metrics.box.recall}")
```

### Using CLI:
```bash
# Validate the model with different confidence thresholds
for conf in 0.1 0.25 0.5 0.75; do
    yolo val model=yolov8n.pt conf=$conf
done
```

These commands will allow you to see how Precision and Recall change with different confidence thresholds. If you need more detailed guidance, please refer to our [Model Validation Documentation](https://docs.ultralytics.com/modes/val/).

If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. Additionally, providing a minimum reproducible example can greatly help in diagnosing any problems. You can find more information on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions! 😊

**Reply by Stephen-Hao** [reply]
Yes! I see, Thank you for your reply!

**Reply by glenn-jocher** [reply]
@Stephen-Hao you're welcome! 😊

To address your query about setting different confidence thresholds for calculating Precision and Recall during model validation in YOLOv8, you can indeed customize the confidence threshold. This allows you to observe how Precision and Recall metrics vary with different thresholds.

### Using Python:
```python
from ultralytics import YOLO

# Load your model
model = YOLO("yolov8n.pt")  # or your custom model path

# Validate the model with different confidence thresholds
for conf_threshold in [0.1, 0.25, 0.5, 0.75]:
    metrics = model.val(conf=conf_threshold)
    print(f"Confidence Threshold: {conf_threshold}")
    print(f"Precision: {metrics.box.precision}")
    print(f"Recall: {metrics.box.recall}")
```

### Using CLI:
```bash
# Validate the model with different confidence thresholds
for conf in 0.1 0.25 0.5 0.75; do
    yolo val model=yolov8n.pt conf=$conf
done
```

These commands will help you see how Precision and Recall change with different confidence thresholds. For more detailed guidance, you can refer to our [Model Validation Documentation](https://docs.ultralytics.com/modes/val/).

If you encounter any issues, please ensure you are using the latest version of the Ultralytics package. Providing a minimum reproducible example can also greatly help in diagnosing any problems. You can find more information on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions! 😊

**Comment by Cyeden** [comment]
请问如果验针集是已经被多尺度分割了，那么再进行验证的时候imgsz应该设置为不同尺度验证的逻辑吗？例如：
if __name__ == '__main__':
    model = YOLO(r'C:\Users\Lazy\Desktop\yolo_detect\weights\best.pt')
  
    multi_scale_sizes = [640, 800, 1024, 1280] 

    for size in multi_scale_sizes:
        model.val(
            data='ultralytics/cfg/datasets/DOTAv1.yaml',
            split='val',
            imgsz=size,
            batch=16,
            project='runs/val',
            name=f'v8n-obb-LSKNet_prev8x-obbpt_{size}',
        )

**Reply by glenn-jocher** [reply]
@Cyeden for multi-scale validation, you can indeed set `imgsz` to different scales within a loop. This approach allows you to evaluate the model's performance across various image sizes. Your implementation looks correct. Ensure you are using the latest version of the Ultralytics package to avoid any potential issues. For more details, you can refer to the [validation documentation](https://docs.ultralytics.com/modes/val/).

**Comment by hefangnan** [comment]
现在我有模型在某个验证集上的所有检测框信息，以txt文件保存起来了，请问我如何不加载模型仅仅通过这些txt文件信息来对这个模型进行评估呢？

**Reply by glenn-jocher** [reply]
@hefangnan thank you for your question! Evaluating a model purely from saved detection information like `.txt` files without loading the original model is currently outside the scope of Ultralytics' built-in tools. However, you can create a custom evaluation script by parsing these files and calculating metrics such as mAP using metrics libraries like [pycocotools](https://pypi.org/project/pycocotools/) or similar. Let us know if you need guidance on YOLO functionalities!

**Comment by jacksun2002** [comment]
您好，验证集输出的json文件里"bbox": [252.898, 399.102, 11.839, 11.781], "score": 0.15335},这些信息是代表着目标框的宽高和置信度大小吗？我想利用目标框的坐标信息，与对应图片的实际标签来计算一些气象方面指标，所以我想确定一下json文件输出信息的确切含义。

**Reply by glenn-jocher** [reply]
@jacksun2002 您好！JSON文件中的`bbox`字段表示检测框的格式为`[x_center, y_center, width, height]`（中心坐标+宽高），`score`是检测结果的置信度分数。这些坐标是绝对像素值，基于验证时设定的`imgsz`参数尺寸。您可以参考[模型验证文档](https://docs.ultralytics.com/modes/val/)中的"保存JSON结果"部分获取更多细节。

**Comment by ihciab** [comment]
请问在模型验证过程中，保存结果到一个目录是必须的吗。例如，我在使用时会多次用到model .val  从而生成过多的目录 即使我将save设置为False

**Reply by glenn-jocher** [reply]
@ihciab hi! Thanks for your question. Saving validation results to a directory is optional. You can control directory creation by setting `project` and `name` arguments to fixed paths, and disable outputs with `plots=False`. For example: 

```python
model.val(project="fixed_dir", name="fixed_subdir", plots=False)
```

This will prevent multiple directories from being generated during repeated validations. See the [Val mode arguments](https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation) for more configuration options. 😊

**Comment by sakoXY** [comment]
您好，请问在执行model.val的过程后，有没有办法保存所有的图片的预测结果呢
```
model = YOLO(model_weight_path)

metrics = model.val(data=val_data_path, plots=True, save_json=True, iou=0.25, project=output_path, batch=16, save_txt=True) 
```
我在文档中只找到了保存json和txt的，如果我希望也能保存所有的图片的预测结果应该怎么做呢？

**Comment by sea190** [comment]
The question about val of yoloworld, and I used my custom dataset, and I make sure the format of ma dataset is corret. While when I run the val.py the code still validate the categories of COCO Dataset instead of my dataset? How can I solve this problem?

**Reply by ihciab** [reply]
You can try specifying the path of the validation set. When calling the val function, specify model.val(data=‘your_data’)

**Reply by glenn-jocher** [reply]
@sea190 for YOLO-World validation with custom datasets, ensure your dataset YAML file correctly defines the class names and paths. YOLO-World may default to COCO categories if the dataset configuration isn't properly specified. Try:

```python
model.val(data='path/to/your/custom_dataset.yaml')
```

Make sure your YAML file includes the correct `names` field with your custom class names and verify the `val` path points to your validation images and labels. If issues persist, check that your dataset follows the expected format for YOLO-World validation.

**Comment by Cherishzyh** [comment]
hello,  I read some dialog about the IOU threshold and confidence threshold and have some questions.

**Reply by Cherishzyh** [reply]
1. the document shows that 'iou' is the threshold for NMS.Is it also a threshold for calculate the confusion matrix? eg. I hope that if the IoU between the prediction and the ground truth is greater than 0.2, it will be considered a true positive (TP). At the same time, does the NMS threshold also have to be 0.2?

**Reply by Cherishzyh** [reply]
2. For confidence threshold, if I set it to 0.1, that mean all bboxes with a score lower than 0.1 will be removed when calculate the PR-curve?. If so, does it will also be remove in result json or txt ？

**Reply by glenn-jocher** [reply]
@Cherishzyh great questions:

1) The `iou` argument in `val()` controls NMS overlap removal only; it does not change the IoU used to determine TP/FP or mAP. Metrics are computed over fixed IoU thresholds 0.50–0.95. If you need evaluation at IoU=0.2, export predictions and compute custom metrics offline. See the [Performance Metrics deep dive](https://docs.ultralytics.com/guides/yolo-performance-metrics/).

2) The `conf` argument is a minimum score filter. Setting `conf=0.1` removes all predictions with score < 0.1 prior to evaluation and from saved outputs (`save_json`/`save_txt`). You can still access PR data from the results object:
```python
from ultralytics import YOLO
model = YOLO("yolo11n.pt")
res = model.val(conf=0.1, iou=0.7, save_json=True, save_txt=True, plots=True)
print(res.box.p_curve)  # precision curve vs score thresholds
print(res.box.r_curve)  # recall curve vs score thresholds
```
Argument details are in the [Val mode arguments](https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation).

**Comment by liujixingit** [comment]
YOLOv8 调用验证功能开启save_txt后保存的坐标如何还原到原图，我尝试了直接×原图宽高，还原的坐标是错误的。save_txt保存的结果是相对什么计算得出的归一化坐标？

**Comment by carlos-cell** [comment]
请问，如何输出不同尺度的mAP值呢，即APs、APm、APl

**Comment by AA278-HUB** [comment]
I have a model trained on a dataset containing four classes (car, bus, motorbike, and trunk). Now, I have a new dataset with only one class (vehicle). How can I validate this model using the new dataset, given that I only care about whether the model can identify the vehicle and its position, without concern for the specific type of vehicle?

**Reply by AA278-HUB** [reply]
f I want to change the validation rules in this specific situation, which .py files need to be modified

**Reply by glenn-jocher** [reply]
@liujixingit `save_txt` writes normalized YOLO `xywh` (center x/y, width, height) relative to the original image size; recover pixel `xyxy` via the original `W,H` then center→corner conversion, e.g.:
```python
# line format: class x y w h [conf]
cls, x, y, w, h, *rest = map(float, line.split())
x1 = (x - w/2) * W; y1 = (y - h/2) * H
x2 = (x + w/2) * W; y2 = (y + h/2) * H
```
Function details are in the [save_one_txt reference](https://docs.ultralytics.com/reference/models/yolo/detect/val/#save_one_txt) and the `save_txt` format is noted under [Validation settings](https://docs.ultralytics.com/usage/cfg/#validation-settings).

@carlos-cell To get APs/APm/APl, run COCO-style evaluation with JSON saving and check the printed COCO summary:
```bash
yolo detect val model=yolo11n.pt data=coco.yaml save_json=True
```
This triggers COCO eval inside validation; see [COCO evaluation in val](https://docs.ultralytics.com/reference/models/yolo/detect/val/#coco_evaluate) and the size-specific metrics overview in the [Performance metrics deep dive](https://docs.ultralytics.com/guides/yolo-performance-metrics/).

@AA278-HUB For “any vehicle” validation, use single-class evaluation so all types collapse to one:
```python
metrics = model.val(data='vehicle.yaml', single_cls=True, agnostic_nms=True)
```
Argument behavior is documented in the [Val mode arguments](https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation).

@AA278-HUB You don’t need code changes for the above, but if you must customize matching/metrics, start from `DetectionValidator` and the `val()` entry point: see the [DetectionValidator reference](https://docs.ultralytics.com/reference/models/yolo/detect/val/) and the [`val()` method in engine/model.py](https://docs.ultralytics.com/reference/engine/model/#val); if proposing a PR, please keep scope minimal and include tests.

**Reply by AA278-HUB** [reply]
thank you,bro

**Reply by glenn-jocher** [reply]
Glad it helped—if this resolves your question, please mark the discussion as answered; if not, share any remaining details here and we’ll help.

**Comment by xyzxyzxyzxyzxz** [comment]
I want to read this page in English. Why it automatically become Chinese even though I click "GB English".

**Reply by glenn-jocher** [reply]
It looks like you’re on the Chinese locale (the URL contains `"/zh/"`). Remove `"/zh/"` to view the English version, or use the top‑right language selector (GB English) and hard‑refresh; also ensure your browser’s auto‑translate is disabled.

**Reply by xyzxyzxyzxyzxz** [reply]
Whether I remove "/zh/" in the URL or click "GB English", it will automatically turn to the Chinese page that is "https://docs.ultralytics.com/zh/modes/val/#how-do-i-save-validation-results-to-a-json-file-in-yolo11". But if I click other language such as Japanese, it will show Japanese correctly. Not only this page, other pages have the same problem. And turning on incognito mode don't work either.

**Reply by glenn-jocher** [reply]
Thanks for the detailed description—that behavior sounds like a docs locale‑routing bug on our side rather than anything you can fix locally; we’ll get it forwarded to the web/docs team, and in the meantime could you confirm whether going directly to the English URL [Model Validation with Ultralytics YOLO (English)](https://docs.ultralytics.com/modes/val/) still forces you back to Chinese and share your OS + browser version so we can reproduce it?

**Reply by xyzxyzxyzxyzxz** [reply]
Yes. It will go to Chinese page "https://docs.ultralytics.com/zh/modes/val/" when I click the link you give. My OS is Windows 11 Professional Edition, 25H2, 26200.7171. And my browser is Chrome 142.0.7444.176. My system language is Chinese.

**Reply by glenn-jocher** [reply]
Thanks for the extra details—that confirms this is an issue on the docs side, likely tied to your system/browser language being Chinese and our locale detection forcing `/zh/` even when you pick English. I’ve passed this to the web/docs team to investigate; as a temporary workaround you can try using a browser whose interface language is set to English (or move English to the top of Chrome’s language preferences) so the site stops auto‑redirecting to the Chinese locale.
